vdr/doc-updater.go

```diff
@@ -0,0 +1,125 @@
+	keyCreator nutsCrypto.KeyCreator
+}
+
+// newNamingFnForExistingDID returns a KIDNamingFunc that can be used as param in the keyCreator.New function.
+// It wraps the KIDNamingFunc with the context of the DID of the document.
+// It returns a keyID in the form of the documents DID with the new keys thumbprint as fragment.
+func newNamingFnForExistingDID(existingDID did.DID) nutsCrypto.KIDNamingFunc {
+	return func(pKey crypto.PublicKey) (string, error) {
+		ecPKey, ok := pKey.(*ecdsa.PublicKey)
+		if !ok {
+			return "", errors.New("could not generate kid: invalid key type")
+		}
+
+		if ecPKey.Curve == nil {
+			return "", errors.New("could not generate kid: empty key curve")
+		}
+
+		jwKey, err := jwk.New(pKey)
+		if err != nil {
+			return "", err
+		}
+		err = jwk.AssignKeyID(jwKey)
+		if err != nil {
+			return "", err
+		}
+
+		existingDID.Fragment = jwKey.KeyID()
+
+		return existingDID.String(), nil
+	}
+}
+
+// RotateAuthenticationKey will create a new AuthenticationMethod with the document id as prefix
+// It removes the old authenticationMethod from the document indicated with methodID
+// It adds the new authenticationMethod to the document
+// It requires the methodID to be part of the authenticationMethods
+// FIXME:This method is a bit too high level and should be moved as part of this issue:
+// https://github.com/nuts-foundation/nuts-node/issues/123
+func (u NutsDocUpdater) RotateAuthenticationKey(methodID did.DID, doc *did.Document) error {
+	if err := u.RemoveVerificationMethod(methodID, doc); err != nil {
+		return err
+	}
+
+	if err := u.CreateNewAuthenticationMethodForDocument(doc); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateNewAuthenticationMethodForDocument creates a new VerificationMethod of type JsonWebKey2020 with a freshly generated key
+// and adds it to the provided document
+// FIXME:This method is a bit too high level and should be moved as part of this issue:
+// https://github.com/nuts-foundation/nuts-node/issues/123
+func (u NutsDocUpdater) CreateNewAuthenticationMethodForDocument(doc *did.Document) error {
+	key, keyIDStr, err := u.keyCreator.New(newNamingFnForExistingDID(doc.ID))
+	if err != nil {
+		return err
+	}
+	keyID, err := did.ParseDID(keyIDStr)
+	if err != nil {
+		return err
+	}
+	method, err := did.NewVerificationMethod(*keyID, did.JsonWebKey2020, did.DID{}, key)
+	if err != nil {
+		return err
+	}
+	doc.AddAuthenticationMethod(method)
+	return nil
+}
+
+// getVerificationMethodDiff is a helper function that makes a diff of verificationMethods between
+// a provided current and a proposedDocument. It returns a list with new and removed verificationMethods
+func getVerificationMethodDiff(currentDocument, proposedDocument did.Document) (new, removed []*did.VerificationMethod) {
+	for _, vmp := range proposedDocument.VerificationMethod {
+		for _, mpc := range currentDocument.VerificationMethod {
+			if vmp.ID.Equals(mpc.ID) {
+				goto nextProposedMethod
```

Comment:
Although I used `goto` as well (in BFSWalker), I find this hard to read. Especially because the label which is jumped to is at the end of the for loop. Wouldn't it be better to use `found` bool in combination with a break/continue instead?